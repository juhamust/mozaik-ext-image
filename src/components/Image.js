import React, { Component } from 'react'
import PropTypes from 'prop-types'
import parseUrl from 'url-parse'
import styled from 'styled-components'
import FileImageO from 'react-icons/lib/fa/file-image-o'
import { Widget, WidgetHeader, WidgetBody } from '@mozaik/ui'

let intervalHandle

const ImageWrapper = styled.div``

const ImageCanvas = styled.div`
    width: 100%;
    height: 100%;
`

class Image extends Component {
    constructor(props) {
        super(props)
        this.state = {
            currentUrl: null,
            prevUrl: null,
            counter: 0,
        }
    }

    componentWillUnmount() {
        if (intervalHandle) {
            clearInterval(intervalHandle)
        }
    }

    componentDidMount() {
        let counter = 0
        const parsedUrl = parseUrl(this.props.url)

        parsedUrl.query = parsedUrl.query || {}
        parsedUrl.search = undefined
        parsedUrl.query.counter = parsedUrl.query.counter || this.state.counter

        // In a case no interval defined, just set static URL
        const refreshInterval = parseInt(this.props.refreshInterval, 10)

        // No refresh
        if (!this.props.refreshInterval) {
            this.setState({
                currentUrl: parsedUrl.toString(),
                prevUrl: parsedUrl.toString(),
            })
            return
        }

        // Set new update interval while removing existing
        if (intervalHandle) {
            clearInterval(intervalHandle)
        }

        intervalHandle = setInterval(() => {
            const parsedUrlNew = parseUrl(this.props.url, true)
            counter = parseInt(this.state.counter, 10) + 1
            parsedUrlNew.query.counter = counter
            this.setState({
                counter: counter,
                currentUrl: parsedUrlNew.toString(),
            })
        }, refreshInterval)
    }

    afterImageLoad() {
        // No-op
    }

    render() {
        const { backgroundColor, backgroundPosition, backgroundSize } = this.props
        const url = this.state.currentUrl || this.props.url
        const prevUrl = this.state.prevUrl || this.state.currentUrl || this.props.url

        const prevStyle = {
            backgroundImage: `url(${prevUrl})`,
            backgroundSize,
            backgroundColor,
        }
        const divStyle = {
            backgroundImage: `url(${url})`,
            backgroundSize,
            backgroundColor,
            backgroundPosition,
        }
        const wrapStyle = Object.assign({}, this.props.wrapStyle, {
            height: '100%',
        })
        const imgStyle = {
            display: 'none',
        }

        // If refreshing is in use, do double-buffering
        let imageArea = <ImageCanvas style={divStyle} />
        if (this.props.refreshInterval) {
            imageArea = (
                <ImageWrapper style={wrapStyle}>
                    <img src={prevUrl} style={imgStyle} onLoad={this.afterImageLoad()} />
                    <ImageCanvas style={prevStyle}>
                        <ImageCanvas style={divStyle} />
                    </ImageCanvas>
                </ImageWrapper>
            )
        }

        return (
            <Widget>
                {this.props.title && (
                    <WidgetHeader
                        title={this.props.title}
                        subject="Status"
                        subjectPlacement="append"
                        icon={FileImageO}
                    />
                )}
                <WidgetBody>{imageArea}</WidgetBody>
            </Widget>
        )
    }
}

Image.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string,
    refreshInterval: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    backgroundSize: PropTypes.string,
    backgroundColor: PropTypes.string,
    backgroundPosition: PropTypes.string,
    wrapStyle: PropTypes.object,
}

Image.defaultProps = {
    refreshInterval: 5000,
    backgroundSize: 'cover',
    backgroundPosition: 'center center',
}

export default Image
