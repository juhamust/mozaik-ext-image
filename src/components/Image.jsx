/* jshint esnext: true */
var _ = require('lodash');
var urlib = require('url');
var React  = require('react');
var classSet = require('react-classset');

var intervalHandle;

var Image = React.createClass({

  getInitialState() {
    return {
      url: null,
      prevUrl: null,
      counter: 0
    };
  },

  componentWillUnmount() {
    if (intervalHandle) {
      clearInterval(intervalHandle);
    }
  },

  componentDidMount() {
    var counter;
    var url = urlib.parse(this.props.url, true);


    url.query = url.query || {};
    url.search = undefined;
    url.query.counter = url.query.counter || this.state.counter;

    // In a case no interval defined, just set static URL
    if (!this.props.refreshInterval) {
      this.setState({
        url: url,
        prevUrl: url
      });
      return;
    }

    if(intervalHandle) {
      clearInterval(intervalHandle);
    }

    this.props.refreshInterval = parseInt(this.props.refreshInterval, 10);

    intervalHandle = setInterval(() => {
      var prevUrl = _.cloneDeep(url);
      counter = parseInt(this.state.counter, 10) + 1;
      url.query.counter = counter.toString();
      //console.log('prev', prevUrl.query.counter, ' - url', url.query.counter);

      this.setState({
        counter: counter,
        prevUrl: prevUrl,
        url: url
      });
    }, (this.props.refreshInterval * 1000));
  },

  afterImageLoad() {
    // No-op
  },

  render() {
    var url = urlib.format(this.state.url);
    var prevUrl = urlib.format(this.state.prevUrl);

    var prevStyle = {
      backgroundImage: 'url(' + prevUrl + ')',
      backgroundSize: this.props.backgroundSize || 'cover',
      backgroundColor: this.props.backgroundColor
    };
    var divStyle = {
      backgroundImage: 'url(' + url + ')',
      backgroundSize: this.props.backgroundSize || 'cover',
      backgroundColor: this.props.backgroundColor
    };
    var wrapStyle = _.defaults((this.props.wrapStyle || {}), {
      height: '100%'
    });
    var imgStyle = {
      display: 'none'
    };

    // If refreshing is in use, do double-buffering
    var imageArea = (<div className="image__background" style={divStyle}></div>);
    if (this.props.refreshInterval) {
      imageArea = (<div style={wrapStyle}>
        <img src={prevUrl} style={imgStyle} onload={this.afterImageLoad()} />
        <div className="image__background" style={prevStyle}>
          <div className="image__background" style={divStyle}></div>
        </div>
      </div>);
    }

    return (
      <div>
        <div className="widget__header">
          {this.props.title || this.props.url}
          <i className="fa fa-picture-o" />
        </div>
        <div className="widget__body">
          {imageArea}
        </div>
      </div>
    );
  }

});

module.exports = Image;