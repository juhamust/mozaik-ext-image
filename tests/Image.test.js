import React from 'react'
import Image from '../src/components/Image'

it('matches with snapshot', () => {
    const wrapper = shallow(<Image url="http://domain.com/image.png" />)
    expect(wrapper).toMatchSnapshot()
})

it('has optional title', () => {
    const wrapper = shallow(<Image title="My title" url="http://domain.com/image.png" />)
    expect(wrapper).toMatchSnapshot()
})

it('updates image', () => {
    return new Promise((resolve, reject) => {
        const wrapper = shallow(
            <Image
                refreshInterval={100}
                title="My title"
                url="http://domain.com/image.png?foo=bar"
            />
        )

        const validate = () => {
            try {
                expect(wrapper.find('img').prop('src')).toEqual(
                    'http://domain.com/image.png?foo=bar&counter=2'
                )
            } catch (error) {
                reject(error)
            }
        }

        setTimeout(() => {
            wrapper.update()
            validate()
            resolve()
        }, 300)
    })
})
