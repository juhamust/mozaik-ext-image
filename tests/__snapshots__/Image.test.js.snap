// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c0 {
  width: 100%;
  height: 100%;
}

<Image
  backgroundPosition="center center"
  backgroundSize="cover"
  refreshInterval="5000"
  url="http://domain.com/image.png"
>
  <div>
    <div
      className="widget__body"
    >
      <styled.div
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <div
          className=""
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <img
            src="http://domain.com/image.png"
            style={
              Object {
                "display": "none",
              }
            }
          />
          <styled.div
            style={
              Object {
                "backgroundColor": undefined,
                "backgroundImage": "url(http://domain.com/image.png)",
                "backgroundSize": "cover",
              }
            }
          >
            <div
              className="c0"
              style={
                Object {
                  "backgroundColor": undefined,
                  "backgroundImage": "url(http://domain.com/image.png)",
                  "backgroundSize": "cover",
                }
              }
            >
              <styled.div
                style={
                  Object {
                    "backgroundColor": undefined,
                    "backgroundImage": "url(http://domain.com/image.png)",
                    "backgroundPosition": "center center",
                    "backgroundSize": "cover",
                  }
                }
              >
                <div
                  className="c0"
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "backgroundImage": "url(http://domain.com/image.png)",
                      "backgroundPosition": "center center",
                      "backgroundSize": "cover",
                    }
                  }
                />
              </styled.div>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </div>
  </div>
</Image>
`;
